pkg/hooks/manager.go

```diff
@@ -187,54 +187,67 @@ func (m *Manager) OnDefineDomain(domainSpec *virtwrapApi.DomainSpec, vmi *v1.Vir
-			if callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {
-				vmiJSON, err := json.Marshal(vmi)
-				if err != nil {
-					return "", fmt.Errorf("Failed to marshal VMI spec: %v", vmi)
-				}
 
-				conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
-				if err != nil {
-					log.Log.Reason(err).Infof(dialSockErr, callback.SocketPath)
-					return "", err
-				}
-				defer conn.Close()
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.OnDefineDomainHookPointName]
+	if !found {
+		return string(domainSpecXML), nil
+	}
 
-				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
-				defer cancel()
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return "", fmt.Errorf("Failed to marshal VMI spec: %v", vmi)
+	}
 
-				switch callback.Version {
-				case hooksV1alpha1.Version:
-					client := hooksV1alpha1.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha1.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				case hooksV1alpha2.Version:
-					client := hooksV1alpha2.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha2.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				default:
-					panic("Should never happen, version compatibility check is done during Info call")
-				}
-			}
+	for _, callback := range callbacks {
+		var err error
+		domainSpecXML, err = m.onDefineDomainCallback(callback, domainSpecXML, vmiJSON)
+		if err != nil {
+			return "", err
 		}
 	}
+
 	return string(domainSpecXML), nil
 }
 
+func (m *Manager) onDefineDomainCallback(callback *callBackClient, domainSpecXML, vmiJSON []byte) ([]byte, error) {
+	conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
+	if err != nil {
+		log.Log.Reason(err).Infof(dialSockErr, callback.SocketPath)
+		return nil, err
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	switch callback.Version {
+	case hooksV1alpha1.Version:
+		client := hooksV1alpha1.NewCallbacksClient(conn)
+		result, err := client.OnDefineDomain(ctx, &hooksV1alpha1.OnDefineDomainParams{
+			DomainXML: domainSpecXML,
+			Vmi:       vmiJSON,
+		})
+		if err != nil {
+			return nil, err
+		}
+		domainSpecXML = result.GetDomainXML()
+	case hooksV1alpha2.Version:
+		client := hooksV1alpha2.NewCallbacksClient(conn)
+		result, err := client.OnDefineDomain(ctx, &hooksV1alpha2.OnDefineDomainParams{
+			DomainXML: domainSpecXML,
+			Vmi:       vmiJSON,
+		})
+		if err != nil {
+			return nil, err
+		}
+		domainSpecXML = result.GetDomainXML()
+	default:
+		log.Log.Errorf("Unsupported callback version: %s", callback.Version)
```

Comment:
Please split this into 2 commits(1 refactor & 1 changing the panic). It would be more obvious IMHO. 

I am a little bit skeptical if we should ignore the sidecar... Panic is not great but maybe we could gracefully exit as we can't satisfy user expectations?

Reply:
@vasiliy-ul @iholder-redhat could you guys comment on the panic part?

Reply:
As we discussed in https://github.com/kubevirt/kubevirt/pull/7642#discussion_r859665230: originally the panic was unreachable:

> I think this panic was unreachable in the original code since the switch was inside the condition `if 1callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {1`. So yeah, removing it should do no harm.

If it makes things clearer then spliting the cahnge into several commits (with proper reasoning) might be a good idea.

Reply:
Done, split into 2 commits.